src/pepr/operator/controllers/keycloak/authservice/authorization-policy.ts

```diff
@@ -0,0 +1,167 @@
+import { Action, AuthorizationPolicy, RequestAuthentication, UDSPackage } from "../../../crd";
+import { getOwnerRef } from "../../utils";
+import { log } from "./authservice";
+import { Action as AuthServiceAction, AuthServiceEvent } from "./types";
+
+const operationMap: {
+  [AuthServiceAction.Add]: "Apply";
+  [AuthServiceAction.Remove]: "Delete";
+} = {
+  [AuthServiceAction.Add]: "Apply",
+  [AuthServiceAction.Remove]: "Delete",
+};
+
+function authserviceAuthorizationPolicy(
+  labelSelector: { [key: string]: string },
+  name: string,
+  namespace: string,
+): AuthorizationPolicy {
+  return {
+    kind: "AuthorizationPolicy",
+    metadata: {
+      name: `${name}-authservice`,
+      namespace,
+    },
+    spec: {
+      action: Action.Custom,
+      provider: {
+        name: "authservice",
+      },
+      rules: [
+        {
+          when: [
+            {
+              key: "request.headers[authorization]",
+              notValues: ["*"],
+            },
+          ],
+        },
+      ],
+      selector: {
+        matchLabels: labelSelector,
+      },
+    },
+  };
+}
+
+function jwtAuthZAuthorizationPolicy(
+  labelSelector: { [key: string]: string },
+  name: string,
+  namespace: string,
+): AuthorizationPolicy {
+  return {
+    kind: "AuthorizationPolicy",
+    metadata: {
+      name: `${name}-jwt-authz`,
+      namespace,
+    },
+    spec: {
+      selector: {
+        matchLabels: labelSelector,
+      },
+      rules: [
+        {
+          from: [
+            {
+              source: {
+                requestPrincipals: [`https://sso.${UDSConfig.domain}/realms/uds/*`],
+              },
+            },
+          ],
+        },
+      ],
+    },
+  };
+}
+
+function authNRequestAuthentication(
+  labelSelector: { [key: string]: string },
+  name: string,
+  namespace: string,
+): RequestAuthentication {
+  return {
+    kind: "RequestAuthentication",
+    metadata: {
+      name: `${name}-jwt-authn`,
+      namespace,
+    },
+    spec: {
+      jwtRules: [
+        {
+          audiences: [name],
+          forwardOriginalToken: true,
+          issuer: `https://sso.${UDSConfig.domain}/realms/uds`,
+          jwksUri: `http://keycloak-http.keycloak.svc.cluster.local:8080/realms/uds/protocol/openid-connect/certs`,
+        },
+      ],
+      selector: {
+        matchLabels: labelSelector,
+      },
+    },
+  };
+}
+
+async function updatePolicy(
+  event: AuthServiceEvent,
+  labelSelector: { [key: string]: string },
+  pkg: UDSPackage,
+) {
+  // type safe map event to operation (either Apply or Delete)
+  const operation = operationMap[event.action];
+  const namespace = pkg.metadata!.namespace!;
+  const generation = (pkg.metadata?.generation ?? 0).toString();
+  const ownerReferences = getOwnerRef(pkg);
+
+  const updateMetadata = (resource: AuthorizationPolicy | RequestAuthentication) => {
+    resource!.metadata!.name = resource!.metadata!.name + "-" + generation;
```

Comment:
Discussed offline but think we can delete the generation suffix here and simplify the purge logic.